File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var longestIncreasingPath = function(matrix) {
    const m = matrix.length;
    const n = matrix[0].length;
    const memo = Array.from({ length: m }, () => Array(n).fill(0));
    const dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]];
    
    const dfs = (r, c) => {
        if (memo[r][c] !== 0) return memo[r][c];
        
        let maxLen = 1;
        for (const [dr, dc] of dirs) {
            const nr = r + dr;
            const nc = c + dc;
            
            if (nr >= 0 && nr < m && nc >= 0 && nc < n 
                && matrix[nr][nc] > matrix[r][c]) {
                maxLen = Math.max(maxLen, 1 + dfs(nr, nc));
            }
        }
        
        memo[r][c] = maxLen;
        return maxLen;
    };
    
    let result = 0;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            result = Math.max(result, dfs(i, j));
        }
    }
    
    return result;
};

File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.java
class Solution {
    private int[][] memo;
    private int[][] matrix;
    private int m, n;
    private int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    
    public int longestIncreasingPath(int[][] matrix) {
        this.matrix = matrix;
        m = matrix.length;
        n = matrix[0].length;
        memo = new int[m][n];
        
        int result = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                result = Math.max(result, dfs(i, j));
            }
        }
        
        return result;
    }
    
    private int dfs(int r, int c) {
        if (memo[r][c] != 0) return memo[r][c];
        
        int maxLen = 1;
        for (int[] d : dirs) {
            int nr = r + d[0];
            int nc = c + d[1];
            
            if (nr >= 0 && nr < m && nc >= 0 && nc < n 
                && matrix[nr][nc] > matrix[r][c]) {
                maxLen = Math.max(maxLen, 1 + dfs(nr, nc));
            }
        }
        
        memo[r][c] = maxLen;
        return maxLen;
    }
}